### PR TITLE
LIVY-213. Implemented interactive statement results recovery.

### PR DESCRIPTION
**Pull request by tc0312**
- Removed statement results history cache in livy-server.
- Proxy statement results history requests to livy-repl.
- Modified RSC repl API:
  - ReplJobRequest to be synchronus and return the id of new statement.
  - GetReplJobResults to return all or some statements in sorted order. Also return current repl state.
  - Reuse Statement and StatementState classes in server, repl and rsc.
- livy-server polls only repl state but not statement results from repl. Repl's state is also updated on statement result requests.
- New StatementState "waiting" to show a stetement that's submitted to repl but hasn't been executed.
- Send statement results to server in JSON instead of JValue to avoid Kryo issue.

### CODE REVIEW

**Comment by tc0312** [comment]
@jerryshao @zjffdu @vanzin Mind if you review my change?

File: server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala [hunk 1]
```diff
@@ -351,6 +348,21 @@ class InteractiveSession(
...
+          // State is also updated when we get statement results from repl, not just here.
+          setSessionStateFromReplState(client.map(_.getReplState.get()))
+          Thread.sleep(1000)
+        }
+      } catch {
+        case _: InterruptedException =>
+      }
+    }
+  })
+
```
Comment (tc0312) [inline server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala]: No. But I want to do this in a separate change.
See // TODO Replace this with a Rpc call from repl to server.

File: repl/src/main/scala/com/cloudera/livy/repl/Session.scala [hunk 2]
```diff
@@ -51,7 +54,9 @@ class Session(interpreter: Interpreter)
   private implicit val formats = DefaultFormats
 
   private var _state: SessionState = SessionState.NotStarted()
-  private var _history = IndexedSeq[Statement]()
+  private val _statements = mutable.Map[Int, Statement]()
```
Comment (jerryshao) [inline repl/src/main/scala/com/cloudera/livy/repl/Session.scala]: I would suggest to use guava's cache instead of this `Map`, that will be easy for you to later support retention.

File: repl/src/main/scala/com/cloudera/livy/repl/Session.scala [hunk 3]
```diff
@@ -70,29 +75,41 @@ class Session(interpreter: Interpreter)
...
+    _statements.synchronized {
+      _statements(statementId) = new Statement(statementId, StatementState.Waiting, null)
+    }
+    Future {
+      _statements.synchronized {
+        _statements(statementId) = new Statement(statementId, StatementState.Running, null)
+      }
+
+      val statement =
+        new Statement(statementId, StatementState.Available, executeCode(statementId, code))
```
Comment (jerryshao) [inline repl/src/main/scala/com/cloudera/livy/repl/Session.scala]: `executeCode` should also be in synchronized block, otherwise interpreter will potentially has concurrency issue.

File: repl/src/test/scala/com/cloudera/livy/repl/BaseSessionSpec.scala [hunk 4]
```diff
@@ -22,18 +22,28 @@ import scala.concurrent.Await
...
 import com.cloudera.livy.LivyBaseUnitTestSuite
+import com.cloudera.livy.rsc.driver.{Statement, StatementState}
 import com.cloudera.livy.sessions.SessionState
 
 abstract class BaseSessionSpec extends FlatSpec with Matchers with LivyBaseUnitTestSuite {
 
   implicit val formats = DefaultFormats
 
-  def withSession(testCode: Session => Any): Unit = {
+  protected def execute(session: Session, code: String): Statement = {
```
Comment (jerryshao) [inline repl/src/test/scala/com/cloudera/livy/repl/BaseSessionSpec.scala]: Can we change to `withSession(session: Session)(code: => String)`?

File: rsc/src/main/java/com/cloudera/livy/rsc/BaseProtocol.java [hunk 5]
```diff
@@ -171,30 +172,30 @@ public RemoteDriverAddress() {
...
+      this.allResults = false;
+      this.from = from;
+      this.size = size;
     }
 
-    public GetReplJobResult() {
-      this(null);
+    public GetReplJobResults() {
+      this.allResults = true;
+      from = size = null;
```
Comment (jerryshao) [inline rsc/src/main/java/com/cloudera/livy/rsc/BaseProtocol.java]: I would avoid this chain assignment.

File: server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala [hunk 6]
```diff
@@ -486,54 +510,28 @@ class InteractiveSession(
...
+  private def transition(newState: SessionState) = synchronized {
     // When a statement returns an error, the session should transit to error state.
     // If the session crashed because of the error, the session should instead go to dead state.
     // Since these 2 transitions are triggered by different threads, there's a race condition.
     // Make sure we won't transit from dead to error state.
-    if (!_state.isInstanceOf[SessionState.Dead] || !state.isInstanceOf[SessionState.Error]) {
-      debug(s"$this session state change from ${_state} to $state")
-      _state = state
+    val areSameStates = _state.getClass() == newState.getClass()
+    val transitFromInactiveToActive = !_state.isActive && newState.isActive
```
Comment (jerryshao) [inline server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala]: What's the usage of condition here, I don't understand?

File: server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala (hunk 1 above)
Comment (jerryshao) [inline server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala]: I agree with @zjffdu , this will create one thread per session, also do we need to pull the repl state so frequently? Each session will send RPC requests per second, that's a big overhead if there's lots of session.

**Comment by tc0312** [comment]
Sorry for the big PR. This was done way before I submit the last PR. I will submit new work in smaller incremental PR.

File: integration-test/src/test/scala/com/cloudera/livy/test/JobApiIT.scala [hunk 7]
```diff
@@ -76,7 +76,8 @@ class JobApiIT extends BaseIntegrationTestSuite with BeforeAndAfterAll with Logg
       assert(list.total === 1)
       val tempSessionId = list.sessions(0).id
 
-      livyClient.connectSession(tempSessionId).verifySessionIdle()
+      val s = livyClient.connectSession(tempSessionId)
```
Comment (tc0312) [inline integration-test/src/test/scala/com/cloudera/livy/test/JobApiIT.scala]: I thought users shouldn't submit RSC jobs while the session's starting.
I will fix the code to re-enable this use case then revert this change.

File: repl/src/main/scala/com/cloudera/livy/repl/Session.scala (hunk 2 above)
Comment (tc0312) [inline repl/src/main/scala/com/cloudera/livy/repl/Session.scala]: Repl was using Guava and caused a lot of jar collision problems with user jars. We don't want to shade Guava either because it's pretty large.
Mind if we keep this PR solely for recovery and deal with this later?

File: server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala (hunk 1 above)
Comment (tc0312) [inline server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala]: Yup. I don't disagree with that. See // TODO Replace this with a Rpc call from repl to server.
Replacing something timer based to Rpc usually introduces race conditions. It's better if I do it in a separate PR so you can review it separately.

File: server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala (hunk 6 above)
Comment (tc0312) [inline server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala]: Which condition? areSameStates or transitFromInactiveToActive?

File: repl/src/main/scala/com/cloudera/livy/repl/Session.scala (hunk 3 above)
Comment (tc0312) [inline repl/src/main/scala/com/cloudera/livy/repl/Session.scala]: Fixed.

File: repl/src/test/scala/com/cloudera/livy/repl/BaseSessionSpec.scala (hunk 4 above)
Comment (tc0312) [inline repl/src/test/scala/com/cloudera/livy/repl/BaseSessionSpec.scala]: Then wouldn't it create a new session every time when the test executes a statement?

File: rsc/src/main/java/com/cloudera/livy/rsc/BaseProtocol.java (hunk 5 above)
Comment (tc0312) [inline rsc/src/main/java/com/cloudera/livy/rsc/BaseProtocol.java]: Fixed.

File: repl/src/main/scala/com/cloudera/livy/repl/Session.scala (hunk 2 above)
Comment (jerryshao) [inline repl/src/main/scala/com/cloudera/livy/repl/Session.scala]: Java `LinkedHashMap` is another one for you to try. It's fine here, we could improve later.

File: repl/src/test/scala/com/cloudera/livy/repl/BaseSessionSpec.scala (hunk 4 above)
Comment (jerryshao) [inline repl/src/test/scala/com/cloudera/livy/repl/BaseSessionSpec.scala]: No, `session` argument is the one you already created. Changing to curry style makes it more scala-ish, you could further define `val b = withSession(session)` and `b(code)` to simplify.

File: server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala (hunk 6 above)
Comment (jerryshao) [inline server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala]: the latter?

File: server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala (hunk 1 above)
Comment (tc0312) [inline server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala]: Sorry I didn't pay attention to that because I'm gonna replace it with a Rpc callback. Will fix.

File: repl/src/test/scala/com/cloudera/livy/repl/BaseSessionSpec.scala (hunk 4 above)
Comment (tc0312) [inline repl/src/test/scala/com/cloudera/livy/repl/BaseSessionSpec.scala]: code is a string of test code, in Scala/Python/R, which's sent to the interpreter class for testing.

Comment (jerryshao) [inline repl/src/test/scala/com/cloudera/livy/repl/BaseSessionSpec.scala]: You may misunderstood my point, my suggestion is to change to:

```
val statement = execute(session)("blahblah")

or

val execution = execute(session)
val statement = execution("blah")
```

Not semantic changes, only the style changes.

Comment (tc0312) [inline repl/src/test/scala/com/cloudera/livy/repl/BaseSessionSpec.scala]: I don't really think it makes a big difference but I will make the change.

Comment (jerryshao) [inline repl/src/test/scala/com/cloudera/livy/repl/BaseSessionSpec.scala]: Yes, there's no big difference, what I want is to keep the same style as original code `withXXX()` or `withXXX(xxx)(...)`. I don't strongly ask you to change, just a suggestion.

File: repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala [hunk 8]
```diff
@@ -73,19 +63,28 @@ class ReplDriver(conf: SparkConf, livyConf: RSCConf)
...
+  def handle(ctx: ChannelHandlerContext, msg: BaseProtocol.ReplJobRequest): Int = {
+    session.execute(msg.code)
   }
 
-  def handle(ctx: ChannelHandlerContext, msg: BaseProtocol.GetReplJobResult): String = {
-    val result = jobFutures.getOrElse(msg.id, null)
-    Option(result).map { r => compact(render(r)) }.orNull
+  /**
+    * Return statement results. Results are sorted by statement id.
+    */
```
Comment (jerryshao) [inline repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala]: Please change to the javadoc comment style:

```
/**
 *
 */
```

File: repl/src/main/scala/com/cloudera/livy/repl/Session.scala [hunk 9]
```diff
@@ -70,29 +75,41 @@ class Session(interpreter: Interpreter)
 
   def state: SessionState = _state
 
-  def history: IndexedSeq[Statement] = _history
+  def statements: mutable.Map[Int, Statement] = _statements
+
+  def execute(code: String): Int = {
+    val statementId = newStatementId.getAndIncrement()
+    _statements.synchronized {
```
Comment (jerryshao) [inline repl/src/main/scala/com/cloudera/livy/repl/Session.scala]: I'm wondering if we could use `ConcurrentHashMap`, then this synchronization block is not necessary, what do you think?

File: repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala [hunk 10]
```diff
@@ -73,19 +63,28 @@ class ReplDriver(conf: SparkConf, livyConf: RSCConf)
...
+  def handle(ctx: ChannelHandlerContext, msg: BaseProtocol.GetReplJobResults): ReplJobResults = {
+    val stmts = if (msg.allResults) {
+      session.statements.values.toArray
+    } else {
+      assert(msg.from != null)
+      assert(msg.size != null)
+      val until = msg.from + msg.size
+      session.statements.filterKeys(id => id >= msg.from && id < until).values.toArray
```
Comment (jerryshao) [inline repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala]: I think there're two potential issues here:

1. Since `statements` is mutable, here without synchronization, the returned result may be inconsistent.
2. Traverse through the who map will be slower if Livy is running for hours, it would be better to use other data structure like sorted map instead.

Also I'm not sure in which scenario do we require to bulk get the statement results?

File: repl/src/main/scala/com/cloudera/livy/repl/Session.scala [hunk 11]
```diff
@@ -140,7 +157,7 @@ class Session(interpreter: Interpreter)
         (EVALUE -> e.getMessage) ~
         (TRACEBACK -> List())
     }
+
+    compact(render(resultInJson))
```
Comment (jerryshao) [inline repl/src/main/scala/com/cloudera/livy/repl/Session.scala]: What is the advantage of render to json here in driver side?

File: server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala (hunk 1 above)
Comment (tc0312) [inline server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala]: I increased the timeout to 30s. Ideally it should be a LivyConf. But this doesn't really matter since I will submit a following PR to remove this thread.

File: repl/src/main/scala/com/cloudera/livy/repl/Session.scala (hunk 9 above)
Comment (tc0312) [inline repl/src/main/scala/com/cloudera/livy/repl/Session.scala]: Realistically speaking, even with HA, no one's gonna connect thousands of livy-server to the same livy-repl. This sounds like a premature optimization.

File: repl/src/main/scala/com/cloudera/livy/repl/Session.scala (hunk 11 above)
Comment (tc0312) [inline repl/src/main/scala/com/cloudera/livy/repl/Session.scala]: We know for sure Kryo can serialize strings, but might not be able to serialize non POJOs.

File: repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala (hunk 10 above)
Comment (tc0312) [inline repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala]: Added synchronized around functions which will mutate `statements`.
> Also I'm not sure in which scenario do we require to bulk get the statement results?

`statements()` in InteractiveSession. Which's called by `get("/:id/statements")` in `InteractiveSessionServlet`.

Comment (jerryshao) [inline repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala]: How about the 2nd question?

**Comment by tc0312** [comment]
Any feedback for iteration 2?

File: repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala [hunk 12]
```diff
@@ -73,19 +63,29 @@ class ReplDriver(conf: SparkConf, livyConf: RSCConf)
     }
   }
 
-  def handle(ctx: ChannelHandlerContext, msg: BaseProtocol.ReplJobRequest): Unit = {
-    Future {
-      jobFutures(msg.id) = session.execute(msg.code).result
-    }
+  def handle(ctx: ChannelHandlerContext, msg: BaseProtocol.ReplJobRequest): Int = synchronized {
```
Comment (jerryshao) [inline repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala]: I'm not sure here `synchronized` is required? From my understanding it is not necessary, since `session.execute` has already been synchronized.

File: repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala [hunk 13]
```diff
@@ -73,19 +63,29 @@ class ReplDriver(conf: SparkConf, livyConf: RSCConf)
...
   }
 
-  def handle(ctx: ChannelHandlerContext, msg: BaseProtocol.GetReplJobResult): String = {
-    val result = jobFutures.getOrElse(msg.id, null)
-    Option(result).map { r => compact(render(r)) }.orNull
+  /**
+   * Return statement results. Results are sorted by statement id.
+   */
+  def handle(ctx: ChannelHandlerContext, msg: BaseProtocol.GetReplJobResults): ReplJobResults =
+    synchronized {
```
Comment (jerryshao) [inline repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala]: Noted that this `synchronized` has a different locker compared to `session.synchronized`, this actually locks nothing.

File: repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala (hunk 12 above)
Comment (tc0312) [inline repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala]: Yes because `def handle(ctx: ChannelHandlerContext, msg: BaseProtocol.GetReplJobResults)` is locking this class too.

File: repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala (hunk 13 above)
Comment (tc0312) [inline repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala]: If we don't get rid of the lock in `def handle(ctx: ChannelHandlerContext, msg: BaseProtocol.ReplJobRequest)` then it will locks something.

Comment (jerryshao) [inline repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala]: Sorry I cannot get your point, can you please elaborate more?

Comment (tc0312) [inline repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala]: I think I have misunderstood your comment. Can you explain which line is doing session.synchronized?

File: repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala (hunk 12 above)
Comment (tc0312) [inline repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala]: Fixed but keeping session.synchronized here for defensive purpose.
